models/pull.go

```diff
@@ -514,10 +514,20 @@ func (pr *PullRequest) SetMerged() (err error) {
 		return err
 	}
 
+	if count, err := sess.Where("id = ? AND has_merged != ?", pr.ID, true).Count(new(PullRequest)); err != nil {
```

Comment:
Should I tell you that `count` can become obsolete right after returning from the function and we'll still get a race condition? 🤔 

My suggestion for locking the PR for the rest of the transaction would be:
```
affected, err := sess.Cols("id").Where("id = ? AND has_merged = ?", pr.ID, false).Update(pr)
```
In `affected` you'll get `1` if the PR requires merging, and `0` if it's already merged. But you'll also be locking the PR record on the database for the rest of the transaction. Other threads landing at this code will have to wait on that PR until resolved (commit or rollback).

Note that this `UPDATE` trick is required because SQLite doesn't support `SELECT ... FOR UPDATE`.

Reply:
Actually, the `UPDATE` trick should be performed on the issue, because the most important indicator (IsClosed) is there.

Reply:
Yeah this was partially what I meant by it not resolving the race completely.

Now looking at it again, we can do the full lock in this bit of code because it's just database mangling and we've already prepped everything.